src/Periphery/Permit2Proxy.sol

```diff
@@ -0,0 +1,357 @@
+    /// @param _assetId the address of the token to approve
+    /// @param _amount amount of tokens to approve
+    /// @param _nonce the nonce to use
+    /// @param _deadline the expiration deadline
+    function getPermit2MsgHash(
+        bytes calldata _diamondCalldata,
+        address _assetId,
+        uint256 _amount,
+        uint256 _nonce,
+        uint256 _deadline
+    ) external view returns (bytes32 msgHash) {
+        // Token Permissions
+        ISignatureTransfer.TokenPermissions
+            memory tokenPermissions = ISignatureTransfer.TokenPermissions(
+                _assetId,
+                _amount
+            );
+        bytes32 permit = _getTokenPermissionsHash(tokenPermissions);
+
+        // Witness
+        Permit2Proxy.LIFICall memory lifiCall = LIFICall(
+            LIFI_DIAMOND,
+            keccak256(_diamondCalldata)
+        );
+        bytes32 witness = _getWitnessHash(lifiCall);
+
+        // PermitTransferWithWitness
+        msgHash = _getPermitWitnessTransferFromHash(
+            PERMIT2.DOMAIN_SEPARATOR(),
+            permit,
+            address(this),
+            _nonce,
+            _deadline,
+            witness
+        );
+    }
+
+    /// Internal Functions ///
+
+    function _getTokenPermissionsHash(
+        ISignatureTransfer.TokenPermissions memory tokenPermissions
+    ) internal pure returns (bytes32) {
+        return
+            keccak256(
+                abi.encode(
+                    PermitHash._TOKEN_PERMISSIONS_TYPEHASH,
+                    tokenPermissions.token,
+                    tokenPermissions.amount
+                )
+            );
+    }
+
+    function _getWitnessHash(
+        Permit2Proxy.LIFICall memory lifiCall
+    ) internal pure returns (bytes32) {
+        return keccak256(abi.encode(WITNESS_TYPEHASH, lifiCall));
+    }
+
+    function _getPermitWitnessTransferFromHash(
+        bytes32 domainSeparator,
+        bytes32 permit,
+        address spender,
+        uint256 nonce,
+        uint256 deadline,
+        bytes32 witness
+    ) internal view returns (bytes32) {
+        bytes32 dataHash = keccak256(
+            abi.encode(
+                PERMIT_WITH_WITNESS_TYPEHASH,
+                permit,
+                spender,
+                nonce,
+                deadline,
+                witness
+            )
+        );
+
+        return
+            keccak256(abi.encodePacked("\x19\x01", domainSeparator, dataHash));
+    }
+
+    function _executeCalldata(bytes memory diamondCalldata) internal {
+        // call diamond with provided calldata
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory data) = LIFI_DIAMOND.call{
+            value: msg.value
+        }(diamondCalldata);
+        // throw error to make sure tx reverts if low-level call was
+        // unsuccessful
+        if (!success) {
+            revert CallToDiamondFailed(data);
+        }
+    }
```

Comment:
The diamond contract that we are calling will emit an event "LiFiTransferStarted" in case of success. That is sufficient. We do not need to waste more gas for additional events.